camus-kafka-coders/src/main/java/com/linkedin/camus/etl/kafka/coders/AvroMessageDecoder.java

```diff
@@ -0,0 +1,120 @@
+import java.nio.ByteBuffer;
+import java.util.Properties;
+
+import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+
+public class AvroMessageDecoder extends MessageDecoder<byte[], GenericData.Record> {
+  private static final byte MAGIC_BYTE = 0x0;
+  private static final int idSize = 4;
+  private final String SCHEMA_REGISTRY_URL = "schema.registry.url";
+  private final String MAX_SCHEMAS_PER_SUBJECT = "max.schemas.per.subject";
+  private final String DEFAULT_MAX_SCHEMAS_PER_SUBJECT = "1000";
+  private static final Logger logger = Logger.getLogger(AvroMessageDecoder.class);
+  protected DecoderFactory decoderFactory;
+  private SchemaRegistryClient schemaRegistry;
+
+  @Override
+  public void init(Properties props, String topicName) {
+    super.init(props, topicName);
+    decoderFactory = DecoderFactory.get();
+    if (props == null) {
+      throw new IllegalArgumentException("Missing schema registry url!");
+    }
+    String url = props.getProperty(SCHEMA_REGISTRY_URL);
+    if (url == null) {
+      throw new IllegalArgumentException("Missing schema registry url!");
+    }
+    String maxSchemaObject = props.getProperty(
+        MAX_SCHEMAS_PER_SUBJECT, DEFAULT_MAX_SCHEMAS_PER_SUBJECT);
+    schemaRegistry = new CachedSchemaRegistryClient(url, Integer.parseInt(maxSchemaObject));
+  }
+
+  private ByteBuffer getByteBuffer(byte[] payload) {
+    ByteBuffer buffer = ByteBuffer.wrap(payload);
+    byte magic = buffer.get();
+    logger.info("MAGIC BYTE" + magic);
+    if (magic != MAGIC_BYTE) {
+      throw new MessageDecoderException("Unknown magic byte!");
+    }
+    return buffer;
+  }
+
+  private Object deserialize(byte[] payload) throws MessageDecoderException {
+    try {
+      ByteBuffer buffer = getByteBuffer(payload);
+      int id = buffer.getInt();
+      Schema schema = schemaRegistry.getByID(id);
+      logger.info(schema.toString());
+      if (schema == null)
+        throw new IllegalStateException("Unknown schema id: " + id);
+      int length = buffer.limit() - 1 - idSize;
+      if (schema.getType().equals(Schema.Type.BYTES)) {
+        byte[] bytes = new byte[length];
+        buffer.get(bytes, 0, length);
+        return bytes;
+      }
+      int start = buffer.position() + buffer.arrayOffset();
+      DatumReader<Object> reader = new GenericDatumReader<Object>(schema);
+      Object object =
+          reader.read(null, decoderFactory.binaryDecoder(buffer.array(), start, length, null));
+
+      if (schema.getType().equals(Schema.Type.STRING)) {
+        object = ((Utf8) object).toString();
+      }
+      return object;
+    } catch (IOException e) {
+      throw new MessageDecoderException("Error deserializing Avro message", e);
+    }
+  }
+
+
+  public CamusWrapper<Record> decode(byte[] payload) {
+    Object object = deserialize(payload);
+    return new CamusAvroWrapper((Record) object);
+  }
```

Comment:
So Camus only handles record type? If so, it's better to check the type of the object and throw a more meaningful exception.